src/TesApi.Web/BatchScheduler.BatchPools.cs

```diff
@@ -0,0 +1,258 @@
+            var containerInfo = await containerRegistryProvider.GetContainerRegistryInfoAsync(tesTask.Executors.FirstOrDefault()?.Image);
+            var registryServer = containerInfo is null ? default : containerInfo.RegistryServer;
+
+            var vmName = string.IsNullOrWhiteSpace(hostname) ? "<none>" : hostname;
+            var vmSize = virtualMachineInformation.VmSize ?? "<none>";
+            var isPreemptable = virtualMachineInformation.LowPriority;
+            registryServer ??= "<none>";
+            identityResourceId ??= "<none>";
+
+            // Generate hash of everything that differentiates this group of pools
+            var displayName = $"{vmName}:{vmSize}:{isPreemptable}:{registryServer}:{identityResourceId}";
+            var hash = BatchUtils.ConvertToBase32(SHA1.HashData(Encoding.UTF8.GetBytes(displayName))).TrimEnd('='); // This becomes 32 chars
+
+            // Build a PoolName that is of legal length, while exposing the most important metadata without requiring user to find DisplayName
+            // Note that the hash covers all necessary parts to make name unique, so limiting the size of the other parts is not expected to appreciably change the risk of collisions. Those other parts are for convenience
+            var remainingLength = PoolKeyLength - hash.Length - 2; // 50 is max name length, 2 is number of inserted chars. This will always be 16 if we use an entire SHA1
+            var visibleVmSize = LimitVmSize(vmSize, Math.Max(remainingLength - vmName.Length, 6));
+            var visibleHostName = vmName[0..Math.Min(vmName.Length, remainingLength - visibleVmSize.Length)];
+            var name = LimitChars($"{visibleHostName}-{visibleVmSize}-{hash}");
+
+            // Trim DisplayName if needed
+            if (displayName.Length > 1024)
+            {
+                // Remove "path" of identityResourceId
+                displayName = displayName[..^identityResourceId.Length] + identityResourceId[(identityResourceId.LastIndexOf('/') + 1)..];
+                if (displayName.Length > 1024)
+                {
+                    // Trim end, leaving fake elipsys as marker
+                    displayName = displayName[..1021] + "...";
+                }
+            }
+
+            return (name, displayName);
+
+            static string LimitVmSize(string vmSize, int limit)
+            {
+                // First try optimizing by removing "Standard_" prefix.
+                var standard = "Standard_";
+                return vmSize.Length <= limit
+                    ? vmSize
+                    : vmSize.StartsWith(standard, StringComparison.OrdinalIgnoreCase)
+                    ? LimitVmSize(vmSize[standard.Length..], limit)
+                    : vmSize[^limit..];
+            }
+
+            static string LimitChars(string text) // ^[a-zA-Z0-9_-]+$
```

Comment:
Could this be simplified with these built in methods?

char.IsNumber
char.IsLetter

Reply:
char.IsAsciiLetterOrDigit is the right answer. I'll change to it.